docs/dev/design/redirects.rst

```diff
@@ -0,0 +1,338 @@
+
+Sphinx HTML to HTMLDir:
+   Allows to redirect HTML URLs to clean-URLs.
+   Useful in case a project changed the style of their URLs.
+
+   They apply to all projects, not just Sphinx projects.
+
+How other services implement redirects
+--------------------------------------
+
+- Gitbook implementation is very basic,
+  they only allow page redirects.
+
+  https://docs.gitbook.com/integrations/git-sync/content-configuration#redirects
+
+- Cloudflare pages allow to capture placeholders and one wildcard (in any part of the URL).
+  They also allow you to set the status code of the redirect,
+  and redirects can be specific in a ``_redirects`` file.
+
+  https://developers.cloudflare.com/pages/platform/redirects/
+
+  They have a limit of 2100 redirects.
+  In case of multiple matches, the topmost redirect will be used.
+
+- Netlify allows to capture placeholders and a wildcard (only allowed at the end).
+  They also allow you to set the status code of the redirect,
+  and redirects can be specific in a ``_redirects`` file.
+
+  - Forced redirects
+  - Match query arguments
+  - Match by country/language and cookies
+  - Per-domain and protocol redirects
+  - In case of multiple matches, the topmost redirect will be used.
+
+  https://docs.netlify.com/routing/redirects/
+
+Improvements
+------------
+
+General improvements
+~~~~~~~~~~~~~~~~~~~~
+
+The following improvements will be applied to all types of redirects.
+
+- Allow choosing the status code of the redirect.
+  We already have a field for this, but it's not exposed to users.
+- Allow to explicitly define the order of redirects.
+  This will be similar to the automation rules feature,
+  where users can reorder the rules so the most specific ones are first.
+  We currently rely on the implicit order of the redirects (updated_at).
+- Allow to disable redirects.
+  It's useful when testing redirects, or when debugging a problem.
+  Instead of having to re-create the redirect,
+  we can just disable it and re-enable it later.
+- Allow to add a short description.
+  It's useful to document why the redirect was created.
+
+Don't run redirects on domains from pull request previews
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We currently run redirects on domains from pull request previews,
+this is a problem when moving a whole project to a new domain.
+
+We don't the need to run redirects on external domains, they
+should be treated as temporary domains.
+
+Normalize paths with trailing slashes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Currently, if users want to redirect a path with a trailing slash and without it,
+they need to create two separate redirects (``/page/`` and ``/page``).
+
+We can simplify this by normalizing the path before matching it.
+
+For example:
+
+From:
+  ``/page/``
+To:
+  ``/new/page``
+
+The from path will be normalized to ``/page``,
+and the filename to match will also be normalized before matching it.
```

Comment:
When normalizing (changing what the user has input), I'd love to be pretty explicit at this. Maybe just raise a validation error at forma level saying it can't has a trailing slash.

If we don't want to go that way, we can change the value before saving it on the database. I'd would avoid to accept `/page/` and save `/page/` in the db but behave as `/page`. This will end being pretty confusing to the user that will try another redirect with just `/page` and to ourselves as well in the end.

Reply:
I'm fine with this.